ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/validation/RequestValidator.java

```diff
@@ -53,81 +56,99 @@
-   *                                snapshot returned will be owned by this class and changes
-   *                                to the snapshot should not affect the source and vice versa
    * @param serviceContext          the {@link ServiceContext} to use
    * @param ksqlConfig              the {@link KsqlConfig} to validate against
    */
   public RequestValidator(
       final Map<Class<? extends Statement>, StatementValidator<?>> customValidators,
       final BiFunction<KsqlExecutionContext, ServiceContext, Injector> injectorFactory,
-      final Supplier<KsqlExecutionContext> snapshotSupplier,
       final ServiceContext serviceContext,
       final KsqlConfig ksqlConfig
   ) {
     this.customValidators = requireNonNull(customValidators, "customValidators");
     this.injectorFactory = requireNonNull(injectorFactory, "injectorFactory");
-    this.snapshotSupplier = requireNonNull(snapshotSupplier, "snapshotSupplier");
     this.serviceContext = requireSandbox(requireNonNull(serviceContext, "serviceContext"));
     this.ksqlConfig = requireNonNull(ksqlConfig, "ksqlConfig");
   }
 
   /**
    * Validates the messages against a snapshot in time of the KSQL engine.
    *
+   * @param sandbox             a sandbox engine to validate against
    * @param statements          the list of statements to validate
    * @param propertyOverrides   a map of properties to override for this validation
    * @param sql                 the sql that generated the list of statements, used for
    *                            generating more useful debugging information
    *
-   * @return the number of new persistent queries that would be created by {@code statements}
+   * @return a list of {@link ValidatedStatement} that represent the result of validation
+   *
    * @throws KsqlException if any of the statements cannot be validated, or the number
    *                       of requested statements would cause the execution context
    *                       to exceed the number of persistent queries that it was configured
    *                       to support
    */
-  public int validate(
+  public List<ValidatedStatement> validate(
+      final KsqlExecutionContext sandbox,
       final List<ParsedStatement> statements,
       final Map<String, Object> propertyOverrides,
       final String sql
   ) {
     validateOverriddenConfigProperties(propertyOverrides);
-    final KsqlExecutionContext ctx = requireSandbox(snapshotSupplier.get());
+    final KsqlExecutionContext ctx = requireSandbox(sandbox);
     final Injector injector = injectorFactory.apply(ctx, serviceContext);
 
-    int numPersistentQueries = 0;
+    final List<ValidatedStatement> validatedStatements = new ArrayList<>();
     for (ParsedStatement parsed : statements) {
       final PreparedStatement<?> prepared = ctx.prepare(parsed);
+
+      // SetProperty and UnsetProperty calls are executed against the
```

Comment:
This is, as I'm sure you're aware, a little nasty... hummmmm....

I'm assuming these are handled by the engine because they are common to standalone, interactive and embedded mode, so it doesn't make sense to explicitly handle them in the rest-api, (as we do for other statements that don't change engine state).

Options I see are:
 - don't handle them in the engine, but explicitly in each of the three locations, (embeded, interactive and standalone).  (SetProperty would then not need to implement `ExecutableDdlStatement` and `SetPropertyCommand` could be deleted).
 - Keep the above code generic and instead keep the special casing to within the `PropertyValidator` and `PropertyExecutor` classes.

Reply:
> don't handle them in the engine, but explicitly in each of the three locations, (embeded, interactive and standalone). (SetProperty would then not need to implement ExecutableDdlStatement and SetPropertyCommand could be deleted).

This is what I wanted to do ... and then got lazy. I will submit a pre-PR for this now that you're keeping me an honest dev.

Reply:
https://github.com/confluentinc/ksql/pull/2712